### PR TITLE
Fix for optimization flags in Limbo library

### PR DESCRIPTION
**Pull request by costashatz**
Hello,

First of all, thanks for this great library! We are using your library as one of our main optimization algorithms in our Gaussian Process/Bayesian Optimization **Limbo** library (check [here](http://www.resibots.eu/limbo/)). We are very much concerned by performance and as such we are switching to the newest Eigen (3.3) and we want to activate all the nice new features (see [this website](http://eigen.tuxfamily.org/index.php?title=3.3)). In particular, we are interested in the *Vectorization* enhancements using specific cpu instructions (avx, fma) and in replacing Eigen's algorithms with LAPACKE's (see [here](https://eigen.tuxfamily.org/dox/TopicUsingBlasLapack.html)) when it is necessary (i.e., big matrices).

Long story short, your library compiled with other flags than what we compile ours, produces some weird artifacts (eigen vector misalignments, seg faults etc.). This PR intends to change the compilation flags of your library. As far as I checked it doesn't break any other part of the library.

Let me know what you think!

*I also changed a small thing in the `CMakeLists.txt` that was not working for my setup.*

Best,
Konstantinos

### CODE REVIEW

**Comment by beniz** [comment]
OK thanks for the contribution. If I understand correctly, you are doing these changes with relation to Eigen3.3 to get improved performances, typically those offered by avz and fma flags. Regarding lapacke, is there a way we can activate it when building libcmaes or is it something that is done outside ? Reason I'ma sking is that the most performance improvements we can factor from your experience directly into libcmaes, the best.

**Comment by costashatz** [comment]
> If I understand correctly, you are doing these changes with relation to Eigen3.3 to get improved performances, typically those offered by avz and fma flags.

Using `avx` and `fma`, supposedly Eigen is faster at vectorized computations. This is what this PR does (in addition to adding the `-O3` flag for compiler optimization). In my use cases, I only saw real differences in very big matrices (i.e., more than 1000 dimensions), but since it is slightly better and we can have it for free, why not?

> Regarding lapacke, is there a way we can activate it when building libcmaes or is it something that is done outside ?

Regarding lapacke/blas, you need to do it when building `libcmaes`. There are the following you need to think about:

- It gives speed ups only when dealing with very big matrices (i.e., more than 1000-2000 dimensions), but I am still experimenting with the gains we get from this. I will get back to you when I have a more concise conclusion.
- It would be easy for me to add LAPACKE/BLAS in your `CMakeLists.txt`, but I am not very familiar with Makefiles.

> Reason I'ma sking is that the most performance improvements we can factor from your experience directly into libcmaes, the best.

Regarding performance improvements, I found out that using Intel's [TBB](https://www.threadingbuildingblocks.org) for parallelization has superior performance over OpenMP. I modified a bit your library so that I can easily make new ESOStrategies (check [here](https://github.com/resibots/libcmaes/tree/custom) -- basically I had to make sure that the inheritance works as it should) and implemented one with TBB parallelization (it is not yet in an open-sourced repo) and it was much better (or it's because TBB requires almost no configuration and integrates better in C++11 code). Again, I can easily add this functionality to libcmaes via CMake if you like (we could keep both OpenMP and TBB also)..

**Comment by beniz** [comment]
Let us know via PR what other improvements you have in store :) Thanks again for the performance increase!

File: CMakeLists.txt [hunk 1]
```diff
@@ -82,6 +78,8 @@ if (CMAKE_VERSION VERSION_LESS 3.1 OR NOT USE_COMPILE_FEATURES)
   endif ()
 endif ()
 
+set (CMAKE_CXX_FLAGS "-O3 -g -mavx -mfma ${CMAKE_CXX_FLAGS}")
+
```
Comment (jschueller) [inline CMakeLists.txt]: not portable at all!

File: CMakeLists.txt [hunk 2]
```diff
@@ -58,13 +58,9 @@ check_include_file (sys/types.h HAVE_SYS_TYPES_H)
 check_include_file (sys/stat.h HAVE_SYS_STAT_H)
 
 # eigen
-find_package (Eigen3 NO_MODULE)
-if (EIGEN3_FOUND)
-  include (${EIGEN3_USE_FILE})
-else ()
-  find_package (Eigen3 REQUIRED)
-  include_directories (${EIGEN3_INCLUDE_DIRS})
-endif()
```
Comment (jschueller) [inline CMakeLists.txt]: you're disabling the module features :(

File: CMakeLists.txt (hunk 1 above)
Comment (costashatz) [reply CMakeLists.txt]: You're absolutely right. This is my problem; I didn't think it thoroughly. In our own fork, we already fixed this issue (https://github.com/resibots/libcmaes/tree/fix_flags_native) and I was planning to make a PR once #162 is merged..

File: CMakeLists.txt (hunk 2 above)
Comment (costashatz) [reply CMakeLists.txt]: This was not working (i.e., I could not select the Eigen version that I wanted) in my setups (I tried in 2-3 different) and that's why I removed it. If @beniz agrees, you can make a PR to get it back (does not really hurt my experiments/setups -- I can always just have one Eigen installed or maybe you can give me a small example on how to use it).

Comment (jschueller) [reply CMakeLists.txt]: you should point to the eigen version using the path to its config file

Comment (costashatz) [reply CMakeLists.txt]: How do I point to the config file? And why is it different than setting `-DEIGEN3_INCLUDE_DIR` to your Eigen folder?

Comment (jschueller) [reply CMakeLists.txt]: use Eigen3_DIR to point to the eigen-config.cmake, this does not require the findeigen script anymore.
